File: projekty/PO_Projekt3/lis.py
from zwierze import Zwierze

import random

class Lis(Zwierze):
    def __init__(self, swiat, polozenie):
        super().__init__(swiat, polozenie, 3, 7, 1.00, 1, 0.00, 1.00)

    def akcja(self):
        if not self._zyje:
            return

        if not random.random() < self._szansaRuchu:
            return

        for i in range(self._zasiegRuchu):
            pola = self._swiat.sasiedniePola(self._polozenie)
            pola = [pole for pole in pola if not self._swiat.poleZajete(pole) or self._swiat.organizmNaPolu(pole).getSila() <= self._sila]

            if len(pola) == 0:
                return

            cel = self._swiat.losowePole(pola)

            if self._swiat.poleZajete(cel):
                self.kolizja(self._swiat.organizmNaPolu(cel), True)
                return

            self.przemiescSieNaPole(cel)

    def tegoSamegoTypu(self, inny):
        return isinstance(inny, Lis)

    def stworzDziecko(self, cel):
        return Lis(self._swiat, cel)

    def rysowanie(self):
        return (255, 117, 26)

File: projekty/PO_Projekt2/ProgOb_Projekt2/src/main/java/pl/Dybowski/Adrian/Projekt2/WirtualnySwiat/Zwierzeta/Lis.java
package pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Zwierzeta;

import java.util.HashSet;
import java.util.Iterator;

import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Organizm;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Stale;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Swiat;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Wspolrzedne2D;

public class Lis extends Zwierze {

	public Lis(Swiat swiat, Wspolrzedne2D polozenie) {
		super(swiat, polozenie, Stale.sila.get(Lis.class), Stale.inicjatywa.get(Lis.class),
				Stale.szansaRuchu.get(Lis.class), Stale.zasiegRuchu.get(Lis.class),
				Stale.szansaUcieczki.get(Lis.class), Stale.szansaRozmnazania.get(Lis.class));
	}
	
	@Override
	public void akcja() {
		if (!zyje) return;
		
		if (!swiat.szansaTrafiona(szansaRuchu)) return;
		
		for (int i = 0; i < zasiegRuchu; i++) {
			HashSet<Wspolrzedne2D> cele = swiat.sasiedniePola(polozenie);
			
			Iterator<Wspolrzedne2D> iterator = cele.iterator();
			while (iterator.hasNext()) {
				Wspolrzedne2D polozenie = iterator.next();
				
				if (swiat.poleZajete(polozenie) && swiat.organizmNaPolu(polozenie).getSila() > sila) {
					iterator.remove();
				}
			}
			
			Wspolrzedne2D cel = swiat.losowePole(cele);
			
			if (swiat.poleZajete(cel)) {
				kolizja(swiat.organizmNaPolu(cel), true);
				return;
			}
					
			przemiescSieNaPole(cel);
		}
	}

	@Override
	public boolean tegoSamegoTypu(Organizm inny) {
		return inny instanceof Lis;
	}

	@Override
	public Organizm stworzDziecko(Wspolrzedne2D cel) {
		return new Lis(swiat, cel);
	}
}
